lifecycle-watcher/src/main/java/org/apache/unomi/lifecycle/BundleWatcherImpl.java

```diff
@@ -223,50 +269,109 @@ private void displayLogsForInactiveServices() {
+                }
+                installedFeatures.add(value);
+            } catch (Exception e) {
+                logger.error("Error when installing {} feature", value, e);
+            }
+        });
+        installedFeatures.forEach(value -> featuresToInstall.remove(value));
+    }
+
+    private void startScheduler() {
+        if (scheduledFuture == null || scheduledFuture.isCancelled()) {
+            TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    displayLogsForInactiveBundles(requiredBundles);
+                    displayLogsForInactiveServices();
+                    checkStartupComplete();
+                }
+            };
+            scheduledFuture = scheduler
+                    .scheduleWithFixedDelay(task, checkStartupStateRefreshInterval, checkStartupStateRefreshInterval, TimeUnit.SECONDS);
+        }
+    }
+
+    private void startSchedulerForAdditionalBundles() {
+        if (scheduledFutureAdditionalBundles == null || scheduledFutureAdditionalBundles.isCancelled()) {
+            TimerTask task = new TimerTask() {
+                @Override
+                public void run() {
+                    if (shouldInstallAdditionalFeatures() && !installingFeatureStarted) {
+                        installingFeatureStarted = true;
+                        installFeatures();
+                    }
+                    displayLogsForInactiveBundles(requiredBundlesFromFeatures);
+                    checkStartupComplete();
+                }
+            };
+            scheduledFutureAdditionalBundles = scheduler
+                    .scheduleWithFixedDelay(task, checkStartupStateRefreshInterval, checkStartupStateRefreshInterval, TimeUnit.SECONDS);
+        }
+    }
+
     private void checkStartupComplete() {
         if (!isStartupComplete()) {
-            if (scheduledFuture == null || scheduledFuture.isCancelled()) {
-                TimerTask task = new TimerTask() {
-                    @Override
-                    public void run() {
-                        displayLogsForInactiveBundles();
-                        displayLogsForInactiveServices();
-                        checkStartupComplete();
-                    }
-                };
-                scheduledFuture = scheduler
-                        .scheduleWithFixedDelay(task, checkStartupStateRefreshInterval, checkStartupStateRefreshInterval, TimeUnit.SECONDS);
-            }
+            startScheduler();
             return;
         }
         if (scheduledFuture != null) {
             scheduledFuture.cancel(true);
             scheduledFuture = null;
         }
+
+        if (!allAdditionalBundleStarted()) {
+            startSchedulerForAdditionalBundles();
+            return;
+        }
+        if (scheduledFutureAdditionalBundles != null) {
+            scheduledFutureAdditionalBundles.cancel(true);
+            scheduledFutureAdditionalBundles = null;
+        }
```

Comment:
Since the previous **scheduledFuture** is finished and destroyed we could reuse it instead of adding a new **scheduledFutureAdditionalBundles** variable ?

And the destroy code (cancel and set to null can be extracted to a function called: **destroyScheduler**)